// mongodb-core/index.js
'use strict';

var BSON = require('bson');
var require_optional = require('require_optional');

try {
  // Attempt to grab the native BSON parser
  var BSONNative = require_optional('bson-ext');
  // If we got the native parser, use it instead of the
  // Javascript one
  if (BSONNative) {
    BSON = BSONNative;
  }
} catch (err) {} // eslint-disable-line

module.exports = {
  MongoError: require('mongodb-core/lib/error').MongoError,
  MongoNetworkError: require('mongodb-core/lib/error').MongoNetworkError,
  Connection: require('mongodb-core/lib/connection/connection'),
  Server: require('mongodb-core/lib/topologies/server'),
  ReplSet: require('mongodb-core/lib/topologies/replset'),
  Mongos: require('mongodb-core/lib/topologies/mongos'),
  Logger: require('mongodb-core/lib/connection/logger'),
  Cursor: require('mongodb-core/lib/cursor'),
  ReadPreference: require('mongodb-core/lib/topologies/read_preference'),
  Sessions: require('mongodb-core/lib/sessions'),
  BSON: BSON,
  // Raw operations
  Query: require('mongodb-core/lib/connection/commands').Query,
  // Auth mechanisms
  MongoCR: require('mongodb-core/lib/auth/mongocr'),
  X509: require('mongodb-core/lib/auth/x509'),
  Plain: require('mongodb-core/lib/auth/plain'),
  GSSAPI: require('mongodb-core/lib/auth/gssapi'),
  ScramSHA1: require('mongodb-core/lib/auth/scram'),
  // Utilities
  parseConnectionString: require('mongodb-core/lib/uri_parser')
};
